Pr/fix mother globe and slave globe behaviour mechanics to be more retail accurate

m241dan opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Makes the Slave Glove spawning mechanics and behaviour a little more retail. I will admit that this is not perfect because of the crappiness of "onMobRoam" timing  / mechanics. The slave globes tend to surge instead of follow closely, but it's mostly pretty good.

This is probably "best" until there is a "multiple pet" mechanic.

## Steps to test these changes

1. Start Server
2. Visit MG
3. Watch, observe, chuckle, maybe frown a little at the surging, but be mostly happy with this.

m241dan committed:
[lua] added Mother Globe and Slave Gloves to ID.mob for The Shrine of RuAvitau

Cleaner, and overall more efficient.

m241dan committed:
[lua] added more accurate Mother Globe slave mechanics

This isn't perfect because it uses onRoam so it might not be "exactly" 30s intervals.
- Mother Globe will now spawn while idle.
- Mother Globe will not immediately respawn the 6th when it is killed if past the 30s interval.
- Combat mechanics remain mostly unchanged.

m241dan committed:
[sql] change slave globes to be have ROAMFLAGS_EVENT

m241dan committed:
[lua] added helpers for dealing with converting relative positions of a mob to absolute in world space

m241dan committed:
[lua] added more "retail accurate" spawning positions and following mechanics for slave globes

This isn't perfect, since the "onMobRoam" script is on a 3s delay, the slaves kind of surge
after Mother Globe. However, until someone wants to implement a "multiple pets" mechanic,
this works pretty well.

TeoTwawki commented:
> crappiness of "onMobRoam" timing / mechanics.

yeah somewhere in the future we need some new functionality. Some Goblins/Moblins in oldton have this behavior of follow-the-leader as well

m241dan commented on scripts/zones/The_Shrine_of_RuAvitau/mobs/Mother_Globe.lua:
```diff
@@ -4,40 +4,109 @@
+        local slaveGlobe = GetMobByID(slaveGlobeID)
+
+        if slaveGlobe:isSpawned() then
+            local slavePositionSlot = slaveGlobe:getLocalVar("slavePositionSlot")
```
missed this in my refactor, should be using the calculateSlaveGlobePosition() method

m241dan commented on scripts/zones/The_Shrine_of_RuAvitau/mobs/Mother_Globe.lua:
```diff
@@ -4,40 +4,109 @@
+    for _, slaveGlobeID in ipairs(ID.mob.SLAVE_GLOBES) do
+        local slaveGlobe = GetMobByID(slaveGlobeID)
+
+        spawnCount = spawnCount + 1
```
This should just be the index since it's not --really-- doing what I wanted...

m241dan reviewed (commented):
I'll clean this up at some point.

zach2good commented on scripts/globals/utils.lua:
```diff
@@ -551,3 +551,23 @@ function utils.mobTeleport(mob, hideDuration, pos, disAnim, reapAnim)
+end
+
+
+function utils.lateralTranslateWithOriginRotation(origin, translation)
```
The dangerous thing about this is that it has no awareness of the navmesh, and if the new end-location is valid. We should probably add a binding to lua_zone to ask the navmesh "is the new x, y, z coord valid" for this sort of thing. (Or better yet, move this all down into pathfind/navmesh and make a binding to access it from Lua)

m241dan replied:
Added a raycast check.

I don't think this is strictly what you wanted. But, I think it works. In terms of pulling this stuff fully in to the core, I'll leave that to someone who is more familiar with our overall navmesh / mob pathing situation.

zach2good replied:
See my writeup here for why raycast doesn't solve this issue (and why its generally pretty awful):
https://github.com/LandSandBoat/server/issues/1470

You're better off making a binding for `zone:isValidPosition(x, y, z)` that hooks up to `navmesh::validPosition`

m241dan replied:
Ah, hmmm okay. It didn't seem like that would actually give me what I wanted. It just seemed like it was valid within the zone, but didn't necessarily preclude something on the other side of a wall or "inside" a column?

m241dan replied:
Okay, bound and switched to validPosition. Though, on the Lua side I named it to "isNavigablePoint" because I feel like it's more expressive (see commit msg).

Left the raycast stuff in there but will drop the commit if you want but I figure "more" bindings can't hurt us.

zach2good replied:
I think I'd rather you drop the raycast binding, since it's a fairly well-known term that doesn't operate in the well-known way.

m241dan replied:
removed.

m241dan committed:
[core] add lua zone binding for validPosition called "isNavigablePoint"

The name changing is intentional, "valid position" is too vague in my
opinion, but I'm not going to change the core to match. I feel that
the "isNavigablePoint" is more expressive and conveys better meaning.

m241dan commented:
Done.

On Mon, Jun 20, 2022 at 10:48 AM Zach Toogood ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In scripts/globals/utils.lua
> <https://github.com/LandSandBoat/server/pull/1970#discussion_r901855097>:
>
> > @@ -551,3 +551,23 @@ function utils.mobTeleport(mob, hideDuration, pos, disAnim, reapAnim)
>          end
>      end)
>  end
> +
> +local ffxiRotConversionFactor = 360.0 / 255.0
> +
> +function utils.ffxiRotToDegrees(ffxiRot)
> +    return ffxiRotConversionFactor * ffxiRot
> +end
> +
> +
> +function utils.lateralTranslateWithOriginRotation(origin, translation)
>
> I think I'd rather you drop the raycast binding, since it's a fairly
> well-known term that doesn't operate in the well-known way.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LandSandBoat/server/pull/1970#discussion_r901855097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3KLWL32TP2OZJV3QN2ZSLVQCOGZANCNFSM5YSHNZFA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

zach2good commented on src/map/lua/lua_zone.cpp:
```diff
@@ -187,6 +187,19 @@ void CLuaZone::reloadNavmesh()
     m_pLuaZone->m_navMesh->reload();
 }
 
+bool CLuaZone::isNavigablePoint(const sol::table& point)
+{
+    position_t position {
+        point["x"].get_or<float>(0),
+        point["y"].get_or<float>(0),
+        point["z"].get_or<float>(0),
+        point["moving"].get_or<uint16>(0),
+        point["rot"].get_or<uint8>(0)
+    };
+
+    return m_pLuaZone->m_navMesh->validPosition(position);
+}
+
```
👍

zach2good commented on scripts/zones/The_Shrine_of_RuAvitau/IDs.lua:
```diff
@@ -35,6 +35,11 @@ zones[xi.zone.THE_SHRINE_OF_RUAVITAU] =
+        MOTHER_GLOBE     = 17506396,
+        SLAVE_GLOBES     =
+        {
+            17506397, 17506398, 17506399, 17506400, 17506401, 17506402
```
You don't need to define more IDs here, when you can know for sure that:
```lua
local SLAVE_GLOBE_1 = ID.mobs.MOTHER_GLOBE + 1
local SLAVE_GLOBE_2 = ID.mobs.MOTHER_GLOBE + 2
local SLAVE_GLOBE_3 = ID.mobs.MOTHER_GLOBE + 3
local SLAVE_GLOBE_4 = ID.mobs.MOTHER_GLOBE + 4
local SLAVE_GLOBE_5 = ID.mobs.MOTHER_GLOBE + 5
local SLAVE_GLOBE_6 = ID.mobs.MOTHER_GLOBE + 6
```

So:
```lua
    for i = 1, 6 do
        local slaveGlobe = GetMobByID(ID.mob.MOTHER_GLOBE + i)
        -- ...
    end
```

m241dan replied:
I'd rather not rely on convention where possible, but I'll make some change.

m241dan replied:
Resolved?

m241dan committed:
[lua] change Slave Globes to be offset from Mother Globe instead of having their items directly in the table
